File: lib/ai-assistants/terminalBoundaries.ts
import type {
  TerminalCommandRule,
  TerminalOperationalBoundaries,
  TerminalCommandListKind,
} from './types';

const NOW = '2026-05-26T00:00:00.000Z';

export const DEFAULT_TERMINAL_WHITELIST: TerminalCommandRule[] = [
  {
    id: 'wl-pwd',
    pattern: 'pwd',
    description: 'Print working directory for project-root verification.',
    category: 'inspection',
    listKind: 'whitelist',
  },
  {
    id: 'wl-git-status',
    pattern: 'git status --short',
    description: 'Read-only repository status snapshot.',
    category: 'version-control',
    listKind: 'whitelist',
  },
  {
    id: 'wl-git-branch',
    pattern: 'git branch --show-current',
    description: 'Resolve active branch without mutating refs.',
    category: 'version-control',
    listKind: 'whitelist',
  },
  {
    id: 'wl-npm-run',
    pattern: 'npm run <script>',
    description: 'Run package scripts defined in package.json (typecheck, test, build).',
    category: 'build',
    listKind: 'whitelist',
  },
  {
    id: 'wl-cargo-check',
    pattern: 'cargo check',
    description: 'Compile-check Rust workspace without producing release artifacts.',
    category: 'build',
    listKind: 'whitelist',
  },
  {
    id: 'wl-rg',
    pattern: 'rg <pattern>',
    description: 'Search codebase with ripgrep; read-only file access.',
    category: 'inspection',
    listKind: 'whitelist',
  },
  {
    id: 'wl-cmux-version',
    pattern: 'cmux --version',
    description: 'Read cmux version for xmux interoperability checks.',
    category: 'inspection',
    listKind: 'whitelist',
  },
  {
    id: 'wl-cmux-workspaces',
    pattern: 'cmux list-workspaces',
    description: 'List cmux workspaces without mutating layout.',
    category: 'inspection',
    listKind: 'whitelist',
  },
];

export const DEFAULT_TERMINAL_BLACKLIST: TerminalCommandRule[] = [
  {
    id: 'bl-rm-rf',
    pattern: 'rm -rf *',
    description: 'Recursive forced deletion of files or directories.',
    category: 'destructive',
    listKind: 'blacklist',
  },
  {
    id: 'bl-sudo',
    pattern: 'sudo *',
    description: 'Privilege escalation or root-context execution.',
    category: 'privilege',
    listKind: 'blacklist',
  },
  {
    id: 'bl-chmod',
    pattern: 'chmod *',
    description: 'Permission mutation on files, directories, or executables.',
    category: 'privilege',
    listKind: 'blacklist',
  },
  {
    id: 'bl-curl-pipe',
    pattern: 'curl * | *',
    description: 'Remote script fetch piped into shell interpreter.',
    category: 'exfiltration',
    listKind: 'blacklist',
  },
  {
    id: 'bl-env-dump',
    pattern: 'env | *',
    description: 'Environment variable dump that may expose secrets.',
    category: 'exfiltration',
    listKind: 'blacklist',
  },
  {
    id: 'bl-ssh-key',
    pattern: 'cat ~/.ssh/*',
    description: 'Read private key material from user home directory.',
    category: 'credential-theft',
    listKind: 'blacklist',
  },
  {
    id: 'bl-git-destructive',
    pattern: 'git push --force',
    description: 'Force-push that can rewrite shared branch history.',
    category: 'destructive',
    listKind: 'blacklist',
  },
  {
    id: 'bl-git-reset-hard',
    pattern: 'git reset --hard',
    description: 'Discard uncommitted work and move HEAD destructively.',
    category: 'destructive',
    listKind: 'blacklist',
  },
];

export interface TerminalEvaluationResult {
  decision: 'allowed' | 'blocked' | 'unknown';
  reason?: string;
  matchedRuleId?: string;
  blockedSegment?: string;
}

export interface TerminalExecSpec {
  command: string;
  args: string[];
}

const EXACT_EXEC: Record<string, TerminalExecSpec> = {
  pwd: { command: 'pwd', args: [] },
  'git status --short': { command: 'git', args: ['status', '--short'] },
  'git branch --show-current': { command: 'git', args: ['branch', '--show-current'] },
  'cargo check': { command: 'cargo', args: ['check'] },
  'cmux --version': { command: 'cmux', args: ['--version'] },
  'cmux list-workspaces': { command: 'cmux', args: ['list-workspaces'] },
};

export function createDefaultTerminalBoundaries(): TerminalOperationalBoundaries {
  return {
    policyMode: 'default-deny',
    whitelist: DEFAULT_TERMINAL_WHITELIST,
    blacklist: DEFAULT_TERMINAL_BLACKLIST,
    updatedAt: NOW,
  };
}

export function normalizeTerminalCommand(command: string): string {
  return command.trim().replace(/\s+/g, ' ');
}

/** Split compound shell strings on ; && || | (no quote-aware parser). */
export function splitCompoundCommand(command: string): string[] {
  const normalized = normalizeTerminalCommand(command);
  if (!normalized) return [];
  return normalized
    .split(/\s*(?:;|&&|\|\||\|)\s*/)
    .map((segment) => segment.trim())
    .filter(Boolean);
}

function matchesPattern(pattern: string, input: string): boolean {
  if (pattern.includes('<')) {
    const prefix = pattern.split('<')[0]?.trim();
    return prefix ? input.startsWith(prefix) : false;
  }
  if (pattern.endsWith(' *')) {
    return input.startsWith(pattern.slice(0, -2));
  }
  return input === pattern || input.startsWith(`${pattern} `);
}

function findMatchingRule(
  input: string,
  rules: TerminalCommandRule[],
): TerminalCommandRule | undefined {
  return rules.find((rule) => matchesPattern(rule.pattern, input));
}

function evaluateSegment(
  segment: string,
  boundaries: TerminalOperationalBoundaries,
  blacklistOnly: boolean,
): TerminalEvaluationResult {
  const normalized = normalizeTerminalCommand(segment);
  if (!normalized) {
    return { decision: 'unknown', reason: 'empty_command' };
  }

  const blackHit = findMatchingRule(normalized, boundaries.blacklist);
  if (blackHit) {
    return {
      decision: 'blocked',
      reason: 'blacklist',
      matchedRuleId: blackHit.id,
      blockedSegment: normalized,
    };
  }

  if (blacklistOnly) {
    return { decision: 'allowed', reason: 'blacklist_only_pass' };
  }

  const whiteHit = findMatchingRule(normalized, boundaries.whitelist);
  if (whiteHit) {
    return {
      decision: 'allowed',
      reason: 'whitelist',
      matchedRuleId: whiteHit.id,
    };
  }

  if (boundaries.policyMode === 'default-deny') {
    return {
      decision: 'blocked',
      reason: 'default_deny',
      blockedSegment: normalized,
    };
  }

  return { decision: 'unknown', reason: 'not_listed', blockedSegment: normalized };
}

export function evaluateTerminalCommandDetailed(
  command: string,
  boundaries: TerminalOperationalBoundaries,
  options?: { blacklistOnly?: boolean },
): TerminalEvaluationResult {
  const blacklistOnly = options?.blacklistOnly === true;
  const segments = splitCompoundCommand(command);
  if (segments.length === 0) {
    return { decision: 'unknown', reason: 'empty_command' };
  }

  for (const segment of segments) {
    const result = evaluateSegment(segment, boundaries, blacklistOnly);
    if (result.decision !== 'allowed') {
      return result;
    }
  }

  return { decision: 'allowed', reason: blacklistOnly ? 'blacklist_only_pass' : 'whitelist' };
}

export function evaluateTerminalCommand(
  command: string,
  boundaries: TerminalOperationalBoundaries,
  options?: { blacklistOnly?: boolean },
): 'allowed' | 'blocked' | 'unknown' {
  return evaluateTerminalCommandDetailed(command, boundaries, options).decision;
}

export function parseAllowedCommandForExec(normalized: string): TerminalExecSpec | null {
  const input = normalizeTerminalCommand(normalized);
  if (EXACT_EXEC[input]) return EXACT_EXEC[input];

  if (input.startsWith('npm run ')) {
    const script = input.slice('npm run '.length).trim();
    if (!script || /\s/.test(script)) return null;
    return { command: 'npm', args: ['run', script] };
  }

  if (input.startsWith('rg ')) {
    const pattern = input.slice(3).trim();
    if (!pattern) return null;
    return { command: 'rg', args: pattern.split(/\s+/) };
  }

  return null;
}

export function rulesFromPatterns(
  patterns: string[],
  listKind: TerminalCommandListKind,
  category = 'custom',
): TerminalCommandRule[] {
  return patterns
    .map((line) => line.trim())
    .filter(Boolean)
    .map((pattern, index) => ({
      id: `${listKind === 'whitelist' ? 'wl' : 'bl'}-custom-${index}-${pattern.slice(0, 12).replace(/\W/g, '')}`,
      pattern,
      description: `Custom ${listKind} rule.`,
      category,
      listKind,
    }));
}

export function mergeBoundaryRules(
  existing: TerminalCommandRule[],
  patterns: string[],
  listKind: TerminalCommandListKind,
): TerminalCommandRule[] {
  const byPattern = new Map(existing.map((rule) => [rule.pattern, rule]));
  return patterns
    .map((line) => line.trim())
    .filter(Boolean)
    .map((pattern, index) => {
      const prior = byPattern.get(pattern);
      if (prior) return prior;
      return {
        id: `${listKind === 'whitelist' ? 'wl' : 'bl'}-custom-${Date.now()}-${index}`,
        pattern,
        description: `Custom ${listKind} rule.`,
        category: 'custom',
        listKind,
      };
    });
}

File: src-tauri/src/terminal_boundaries.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalCommandRule {
    pub id: String,
    pub pattern: String,
    pub description: String,
    pub category: String,
    pub list_kind: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOperationalBoundaries {
    pub policy_mode: String,
    pub whitelist: Vec<TerminalCommandRule>,
    pub blacklist: Vec<TerminalCommandRule>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalEvaluationResult {
    pub decision: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub matched_rule_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blocked_segment: Option<String>,
}

pub fn normalize_terminal_command(command: &str) -> String {
    command.split_whitespace().collect::<Vec<_>>().join(" ")
}

pub fn split_compound_command(command: &str) -> Vec<String> {
    let normalized = normalize_terminal_command(command);
    if normalized.is_empty() {
        return Vec::new();
    }
    normalized
        .split(|c| c == ';')
        .flat_map(|part| part.split("&&"))
        .flat_map(|part| part.split("||"))
        .flat_map(|part| part.split('|'))
        .map(|s| normalize_terminal_command(s))
        .filter(|s| !s.is_empty())
        .collect()
}

fn matches_pattern(pattern: &str, input: &str) -> bool {
    if pattern.contains('<') {
        if let Some(prefix) = pattern.split('<').next() {
            let prefix = prefix.trim();
            return !prefix.is_empty() && input.starts_with(prefix);
        }
        return false;
    }
    if let Some(stripped) = pattern.strip_suffix(" *") {
        return input.starts_with(stripped);
    }
    input == pattern || input.starts_with(&format!("{pattern} "))
}

fn find_matching_rule<'a>(
    input: &str,
    rules: &'a [TerminalCommandRule],
) -> Option<&'a TerminalCommandRule> {
    rules.iter().find(|rule| matches_pattern(&rule.pattern, input))
}

fn evaluate_segment(
    segment: &str,
    boundaries: &TerminalOperationalBoundaries,
    blacklist_only: bool,
) -> TerminalEvaluationResult {
    let normalized = normalize_terminal_command(segment);
    if normalized.is_empty() {
        return TerminalEvaluationResult {
            decision: "unknown".to_string(),
            reason: Some("empty_command".to_string()),
            matched_rule_id: None,
            blocked_segment: None,
        };
    }

    if let Some(rule) = find_matching_rule(&normalized, &boundaries.blacklist) {
        return TerminalEvaluationResult {
            decision: "blocked".to_string(),
            reason: Some("blacklist".to_string()),
            matched_rule_id: Some(rule.id.clone()),
            blocked_segment: Some(normalized),
        };
    }

    if blacklist_only {
        return TerminalEvaluationResult {
            decision: "allowed".to_string(),
            reason: Some("blacklist_only_pass".to_string()),
            matched_rule_id: None,
            blocked_segment: None,
        };
    }

    if let Some(rule) = find_matching_rule(&normalized, &boundaries.whitelist) {
        return TerminalEvaluationResult {
            decision: "allowed".to_string(),
            reason: Some("whitelist".to_string()),
            matched_rule_id: Some(rule.id.clone()),
            blocked_segment: None,
        };
    }

    if boundaries.policy_mode == "default-deny" {
        return TerminalEvaluationResult {
            decision: "blocked".to_string(),
            reason: Some("default_deny".to_string()),
            matched_rule_id: None,
            blocked_segment: Some(normalized),
        };
    }

    TerminalEvaluationResult {
        decision: "unknown".to_string(),
        reason: Some("not_listed".to_string()),
        matched_rule_id: None,
        blocked_segment: Some(normalized),
    }
}

pub fn evaluate_terminal_command_internal(
    command: &str,
    boundaries: &TerminalOperationalBoundaries,
    blacklist_only: bool,
) -> TerminalEvaluationResult {
    let segments = split_compound_command(command);
    if segments.is_empty() {
        return TerminalEvaluationResult {
            decision: "unknown".to_string(),
            reason: Some("empty_command".to_string()),
            matched_rule_id: None,
            blocked_segment: None,
        };
    }

    for segment in segments {
        let result = evaluate_segment(&segment, boundaries, blacklist_only);
        if result.decision != "allowed" {
            return result;
        }
    }

    TerminalEvaluationResult {
        decision: "allowed".to_string(),
        reason: Some(if blacklist_only {
            "blacklist_only_pass".to_string()
        } else {
            "whitelist".to_string()
        }),
        matched_rule_id: None,
        blocked_segment: None,
    }
}

pub fn default_terminal_boundaries() -> TerminalOperationalBoundaries {
    TerminalOperationalBoundaries {
        policy_mode: "default-deny".to_string(),
        updated_at: "2026-05-26T00:00:00.000Z".to_string(),
        whitelist: vec![
            rule("wl-pwd", "pwd", "inspection", "whitelist"),
            rule("wl-git-status", "git status --short", "version-control", "whitelist"),
            rule(
                "wl-git-branch",
                "git branch --show-current",
                "version-control",
                "whitelist",
            ),
            rule("wl-npm-run", "npm run <script>", "build", "whitelist"),
            rule("wl-cargo-check", "cargo check", "build", "whitelist"),
            rule("wl-rg", "rg <pattern>", "inspection", "whitelist"),
        ],
        blacklist: vec![
            rule("bl-rm-rf", "rm -rf *", "destructive", "blacklist"),
            rule("bl-sudo", "sudo *", "privilege", "blacklist"),
            rule("bl-chmod", "chmod *", "privilege", "blacklist"),
            rule("bl-curl-pipe", "curl * | *", "exfiltration", "blacklist"),
            rule("bl-env-dump", "env | *", "exfiltration", "blacklist"),
            rule("bl-ssh-key", "cat ~/.ssh/*", "credential-theft", "blacklist"),
        ],
    }
}

fn rule(id: &str, pattern: &str, category: &str, list_kind: &str) -> TerminalCommandRule {
    TerminalCommandRule {
        id: id.to_string(),
        pattern: pattern.to_string(),
        description: format!("Default {list_kind} rule for {pattern}"),
        category: category.to_string(),
        list_kind: list_kind.to_string(),
    }
}

#[tauri::command]
pub fn evaluate_terminal_command(
    command: String,
    boundaries: Option<TerminalOperationalBoundaries>,
    blacklist_only: Option<bool>,
) -> Result<TerminalEvaluationResult, String> {
    let trimmed = command.trim();
    if trimmed.is_empty() {
        return Err("command must not be empty".to_string());
    }
    let policy = boundaries.unwrap_or_else(default_terminal_boundaries);
    Ok(evaluate_terminal_command_internal(
        trimmed,
        &policy,
        blacklist_only.unwrap_or(false),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn allows_pwd() {
        let boundaries = default_terminal_boundaries();
        let result = evaluate_terminal_command_internal("pwd", &boundaries, false);
        assert_eq!(result.decision, "allowed");
    }

    #[test]
    fn blocks_sudo() {
        let boundaries = default_terminal_boundaries();
        let result = evaluate_terminal_command_internal("sudo ls", &boundaries, false);
        assert_eq!(result.decision, "blocked");
    }

    #[test]
    fn blacklist_only_allows_unknown() {
        let boundaries = default_terminal_boundaries();
        let result = evaluate_terminal_command_internal("claude --help", &boundaries, true);
        assert_eq!(result.decision, "allowed");
    }

    #[test]
    fn blocks_compound_segment() {
        let boundaries = default_terminal_boundaries();
        let result =
            evaluate_terminal_command_internal("git status --short && sudo ls", &boundaries, false);
        assert_eq!(result.decision, "blocked");
    }
}
